File: src/leetcode/java/IslandPerimeter_463.java
package leetcode.java;

public class IslandPerimeter_463 {
  public int islandPerimeter(int[][] grid) {
    int[] per = new int[5];
    int row = grid.length;
    int col = grid[0].length;
    for (int i = 0; i < row; i++) {
      for (int j = 0; j < col; j++) {
        if (grid[i][j] == 1) per[getPerimeter(grid, i, j, row, col)]++;
      }
    }
    int res = 0;
    for (int i = 1; i < per.length; i++) {
      res += i * per[i];
    }
    return res;
  }

  private int getPerimeter(int[][] grid, int i, int j, int row, int col) {
    int perimeter = 0;
    if (i - 1 < 0 || grid[i - 1][j] == 0) perimeter++;
    if (i + 1 >= row || grid[i + 1][j] == 0) perimeter++;
    if (j - 1 < 0 || grid[i][j - 1] == 0) perimeter++;
    if (j + 1 >= col || grid[i][j + 1] == 0) perimeter++;
    return perimeter;
  }
}

File: src/leetcode/python/IslandPerimeter_463.py
class IslandPerimeter_463(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        row, col = len(grid), len(grid[0])
        per = [0] * 5

        def check(k, l):
            permeter = 0
            if i - 1 < 0 or grid[i - 1][j] == 0: permeter += 1
            if i + 1 >= row or grid[i + 1][j] == 0: permeter += 1
            if j - 1 < 0 or grid[i][j - 1] == 0: permeter += 1
            if j + 1 >= col or grid[i][j + 1] == 0: permeter += 1
            return permeter

        for i in range(row):
            for j in range(col):
                if grid[i][j] == 1:
                    per[check(i, j)] += 1

        res = 0
        for i, n in enumerate(per):
            res = i * n
        return res
